examples/end-to-end-examples/support_chatbot/chatbot.py

```diff
@@ -0,0 +1,546 @@
+            self.subchain = self.create_chain(EVAL_30B_TEMPLATE)
+        for sub_QA in all_sub_QA:
+            response = self.subchain(sub_QA[1:])
+            if response['source_documents']:
+                answer = self.clean_response(response['result'].lstrip('\n'))
+                sub_QA_injection += f'Question: {sub_QA[1:]} \nAnswer: {answer}\n'
+
+        SUBQUERY_COMBINE_TEMPLATE = PARTIAL_COMBINE_TEMPLATE.format(sub_QA_injection)
+        combine_chain = self.create_chain(SUBQUERY_COMBINE_TEMPLATE)
+        combine_response = combine_chain(query)
+        combine_answer = self.clean_response(combine_response['result'].lstrip('\n'))
+        sources = ''
+        slack_deduplicate = True
+        for d in combine_response['source_documents']:
+            if d.metadata["score"] < 0.7:
+                if 'message_from_slack' == sources[:18] and slack_deduplicate:
+                    sources = sources + 'slack_data\n'
+                    slack_deduplicate = False
+                else:
+                    sources = sources + f'{d.metadata["file_name"].replace("{slash}", "/")}\n'
+        if not sources:
+            return f'Answer: \n{combine_answer}\n\nIntent: \n{intent_answer}\n\nRelated Sub-questions: \n{sub_QA_injection}'
+        else:
+            return f'Answer: \n{combine_answer}\n\nIntent: \n{intent_answer}\n\nRelated Sub-questions: \n{sub_QA_injection}\nSources: \n{sources}'
+          
+    def chat(self, 
+             query: str) -> str:
+        """Chat with the chatbot given a query
+        
+        Args:
+            query (str): The query to ask the chatbot
+        """
+    
+        if query == "!eval_7b":
+            self.set_eval_state(endpoint_url='https://chatbot-7b-finetuned-rxyfc5.inf.hosted-on.mosaicml.hosting/predict')
+            score = self.evaluate_simple(EVAL_SIMPLE_DIR, EVAL_7B_TEMPLATE)
+            self.reload_chat_state()
+            print(score)
+            return score
+        elif query == "!eval_7b_complex":
+            self.model.endpoint_url = 'https://chatbot-7b-finetuned-rxyfc5.inf.hosted-on.mosaicml.hosting/predict'
+            out = self.evaluate_complex(EVAL_COMPLEX_DIR, EVAL_7B_TEMPLATE)
+            self.model.endpoint_url = self.saved_state['endpoint_url']
+            return out
+        elif query == "!eval_30b":
+            score = self.evaluate_simple(EVAL_SIMPLE_DIR, EVAL_30B_TEMPLATE)
+            print(score)
+            return score
+        elif query == "!eval_30b_complex":
+            out = self.evaluate_complex(EVAL_30B_TEMPLATE, EVAL_30B_TEMPLATE)
+            return out
+        else:
+            # Don't create a new chain on every query
+            if not self.chat_chain:
+                self.chat_chain = self.create_chain(EVAL_30B_TEMPLATE)
+            response = self.chat_chain(query)
+            answer = self.clean_response(response['result'].lstrip('\n'))
+            sources = ''
+            slack_deduplicate = True
```

Comment:
what is the slack dedup arg?

Reply:
time.strftime('message_from_slack_%Y%m%d%H%M%S', time.gmtime(float(timestamp))). I was going to just abstract it away to "slack message", and then make sure "slack message" didn't appear more than once

Reply:
time.strftime('message_from_slack_%Y%m%d%H%M%S', time.gmtime(float(timestamp))) is the a completed thread is saved in oci when emoji is used